cpp/src/Ice/ObjectAdapterFactory.cpp

```diff
@@ -39,59 +37,128 @@ IceInternal::ObjectAdapterFactory::shutdown()
     }
 
     // Deactivate outside the thread synchronization, to avoid deadlocks.
-    for_each(
-        adapters.begin(),
-        adapters.end(),
-        [](const shared_ptr<ObjectAdapterI>& adapter) { adapter->deactivate(); });
+    for (const auto& adapter : adapters)
+    {
+        adapter->deactivate();
+    }
 }
 
 void
-IceInternal::ObjectAdapterFactory::waitForShutdown()
+IceInternal::ObjectAdapterFactory::waitForShutdown() noexcept
 {
     list<shared_ptr<ObjectAdapterI>> adapters;
-
     {
         unique_lock lock(_mutex);
+        if (_shutdownCompleted)
+        {
+            return;
+        }
 
-        //
         // First we wait for the shutdown of the factory itself.
-        //
         _conditionVariable.wait(lock, [this] { return !_instance; });
         adapters = _adapters;
     }
 
     // Now we wait for deactivation of each object adapter.
-    for_each(
-        adapters.begin(),
-        adapters.end(),
-        [](const shared_ptr<ObjectAdapterI>& adapter) { adapter->waitForDeactivate(); });
+    for (const auto& adapter : adapters)
+    {
+        adapter->waitForDeactivate();
+    }
+
+    {
+        lock_guard lock(_mutex);
+        _shutdownCompleted = true;
+    }
+}
+
+void
+IceInternal::ObjectAdapterFactory::waitForShutdownAsync(function<void()> completed) noexcept
+{
+    assert(completed); // the caller (Communicator) makes sure completed is callable.
+
+    bool executeCallback = false;
+    {
+        lock_guard lock(_mutex);
+        if (_shutdownCompleted)
+        {
+            executeCallback = true; // execute outside the lock
+        }
+        else
+        {
+            // Start the shutdown completed thread if there is no callback yet.
+            if (_shutdownCompletedCallbacks.empty())
+            {
+                assert(!_shutdownCompletedThread.joinable());
+                _shutdownCompletedThread = std::thread{[self = shared_from_this()]()
+                                                       {
+                                                           self->waitForShutdown();
+                                                           vector<function<void()>> callbacks;
+                                                           {
+                                                               lock_guard lg(self->_mutex);
+                                                               callbacks = std::move(self->_shutdownCompletedCallbacks);
+                                                           }
+
+                                                           for (const auto& callback : callbacks)
+                                                           {
+                                                               callback();
```

Comment:
We should probably try/catch around these user callbacks.

Reply:
The doc-comment says the callback must not throw any exception.